art/estimators/certification/interval/pytorch.py

```diff
@@ -0,0 +1,376 @@
+
+import logging
+import warnings
+
+import numpy as np
+import torch
+
+from art.estimators.certification.interval.interval import (
+    IntervalConv2D,
+    IntervalDenseLayer,
+    IntervalReLU,
+    IntervalBounds,
+)
+from art.estimators.classification.pytorch import PyTorchClassifier
+
+
+if TYPE_CHECKING:
+    from art.utils import CLIP_VALUES_TYPE, PREPROCESSING_TYPE
+    from art.defences.preprocessor import Preprocessor
+    from art.defences.postprocessor import Postprocessor
+
+
+class ConvertedModel(torch.nn.Module):
+    """
+    Class which converts the supplied pytorch model into an equivalent model
+    which uses abstract operations
+    """
+
+    def __init__(self, model: "torch.nn.Module", channels_first: bool, input_shape: Tuple[int, ...]):
+        super().__init__()
+        modules = []
+        self.interim_shapes: List[Tuple] = []
+        self.device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
+        self.forward_mode: str
+        self.forward_mode = "abstract"
+        self.reshape_op_num = -1
+
+        # pylint: disable=W0613
+        def forward_hook(input_module, hook_input, hook_output):
+            modules.append(input_module)
+            self.interim_shapes.append(tuple(hook_input[0].shape))
+
+        for module in model.children():
+            module.register_forward_hook(forward_hook)
+
+        if channels_first:
+            input_for_hook = torch.rand(input_shape).to(self.device)
+        else:
+            raise ValueError("Please provide data in channels first format")
+
+        input_for_hook = torch.unsqueeze(input_for_hook, dim=0)
+        model(input_for_hook)  # hooks are fired sequentially from model input to the output
+
+        self.ops = torch.nn.ModuleList()
+        for module, shapes in zip(modules, self.interim_shapes):
+            weights_to_supply = None
+            bias_to_supply = None
+            print("registered", type(module))
+            if isinstance(module, torch.nn.modules.conv.Conv2d):
+                if module.weight is not None:
+                    weights_to_supply = torch.tensor(np.copy(module.weight.data.cpu().detach().numpy())).to(self.device)
+                if module.bias is not None:
+                    bias_to_supply = torch.tensor(np.copy(module.bias.data.cpu().detach().numpy())).to(self.device)
+
+                interval_conv = IntervalConv2D(
+                    input_shape=shapes,
+                    in_channels=module.in_channels,
+                    out_channels=module.out_channels,
+                    kernel_size=module.kernel_size,  # type: ignore
+                    stride=module.stride,  # type: ignore
+                    supplied_input_weights=weights_to_supply,
+                    supplied_input_bias=bias_to_supply,
+                    device=self.device,
+                    dilation=module.dilation,  # type: ignore
+                    padding=module.padding,  # type: ignore
```

Comment:
What is the reason for these type ignores?

Reply:
These are ignored due to nn.Conv2d requiring the type hinting indicated in the init of PytorchIntervalConv(torch.nn.Module). However, reading layer information using module.X can return types of a slightly different form (eg Tuple[int, ...] vs Tuple[int]) which nonetheless function when passed into nn.Conv2d. Hence, using an ignore to suppress the warning.